geps/gep-2648/index.md

```diff
@@ -0,0 +1,415 @@
+
+A Direct Policy Attachment is tightly bound to one instance of a particular
+Kind within a single namespace (or to an instance of a single Kind at cluster scope),
+and only modifies the behavior of the object that matches its binding.
+
+As an example, the BackendTLSPolicy is specified in GEP-1897, BackendTLSPolicy -
+Explicit Backend TLS Connection Configuration. This Policy attaches to the
+Service object and tells Gateway API implementations what TLS settings should be
+used to connect to that Service when it is used as a backend by a Route.
+
+See GEP-1897 for all the details of this Policy object.
+
+## Direct Policy design rules
+
+With this example in mind, here are some rules for when a Policy is a Direct
+Policy:
+
+* The number or scope of objects is exactly _one_ object. No label selectors or
+  lists of `targetRef` are allowed.
+* The modifications to be made to the objects don’t have any transitive information -
+  that is, the modifications MUST only affect the single object that the targeted
+  metaresource is bound to, and MUST NOT have ramifications that flow beyond that
+  object. No attaching a Policy to a Gateway and affecting settings in Routes or
+  backends. If a Direct Attached Policy attaches to an object, it can only affect
+  properties _of that object_.
+* In terms of status, it SHOULD be reasonably easy for a user to understand that
+  everything is working - basically, as long as the targeted object exists, and
+  the modifications are valid, the metaresource is valid, and this should be
+  straightforward to communicate in one or two Conditions. The `status` stanza
+  in BackendTLSPolicy is an example of one of the recommended ways to achieve this.
+* Direct Policy Attachment SHOULD only be used to target objects in the same
+  namespace as the Policy object. Allowing cross-namespace references brings in
+  significant security concerns, and/or difficulties about merging cross-namespace
+  policy objects. Notably, Mesh use cases may need to do something like this for
+  consumer policies, but in general, Policy objects that modify the behavior of
+  things outside their own namespace should be avoided unless it uses a handshake
+  of some sort, where the things outside the namespace can opt–out of the behavior.
+  (Notably, this is the design that we used for ReferenceGrant).
+
+## Apply Policies to Sections of a Resource
+
+The `sectionName` field of `targetRef` can be used to target a specific section of other resources:
+
+* Service.Ports.Name
+* Gateway.Listners.Name
+* HTTPRoute.Rules.Name (once they are added in GEP-995, PR at https://github.com/kubernetes-sigs/gateway-api/pull/2593)
+
+For example, the RetryPolicy below applies to a RouteRule inside an HTTPRoute.
+(or rather, it will when GEP-995 merges).
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: HTTPRoute
+metadata:
+  name: http-app-1
+  labels:
+    app: foo
+spec:
+  hostnames:
+  - "foo.com"
+  rules:
+  - name: bar
```

Comment:
This is the sectionName, but I believe the same name can be used in different stanzas that are different objects, which breaks the application of policy to section name.  For example, couldn't I have a rule and a backendRef with the same name?

Reply:
At the present, this is left to the Policy creator to define. I think this is another good follow-up discussion though.

Reply:
The example provided assumes GEP-995 (https://github.com/kubernetes-sigs/gateway-api/pull/2593), I understand.

Reply:
> For example, couldn't I have a rule and a backendRef with the same name?

Hrm, I've definitely never considered this - since the backendRef is itself an object so I've only ever thought of targeting it directly with an policy targetRef to the kind/namespace/name of the ref'd object rather than scoping the targeting to only via it's backend role within a specific HTTPRoute rule.

> I believe the same name can be used in different stanzas that are different objects

For backendRefs yea, you could have two rules targeting the same named backendRef, which I think makes the backendRef name inappropriate as a standalone selector - scoping it _under_ an HTTPRoute rule sectionName (which I believe https://github.com/kubernetes-sigs/gateway-api/pull/2593 suggests making unique when specified) I think would allow precise targeting though, like

```
apiVersion: networking.acme.io/v1alpha2
kind: RetryPolicy
metadata:
  name: foo
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: http-app-1
    sectionName: bar
  maxRetries: 5
  backendRefs:
  - "baz"
```

Reply:
+1 to not using `sectionName` for `backendRef`, and to "section of a HTTPRoute" == a HTTPRouteRule.

Reply:
FWIW the need for "multi-dimensional targeting" isn't unique to BackendTLSPolicy - I hope that expecting all Gateways supporting BackendTLSPolicy to respect a policy applied to a Service for any backendRefs targeting it can be sufficient to start with, but scoping a policy to more limited `from` clauses is also a pattern we'll need to figure out if we want to have AuthorizationPolicy for mesh too (e.g. apply policy to targetRef `destination` and scope ALLOW rules as `from` potentially multiple `origin` sources, similar to the structure of https://istio.io/latest/docs/reference/config/security/authorization-policy/).

Reply:
@robscott 

> I think sectionName would be used exclusively to refer to rule names, and that rule names would need to be unique within a Route when they were specified. 

Ok, I wasn't aware that we had clearly defined `sectionNames` for `HTTPRoute` or any other object.  Where do we document that information and more importantly, how is the constraint enforced?

Reply:
Opened https://github.com/kubernetes-sigs/gateway-api/pull/2946 to improve documentation for `SectionName`